### ISSUE TITLE
[Bug] Data missing from storage after app restart

### ISSUE THREAD

**Issue by andrei-tofan** [ammarahm-ed/react-native-mmkv-storage#195]
**Describe the bug**
We received in the past few months many complaints about users being logged out from the app. After adding logging and investigating the incidents we found out that at random the `getStringAsync` returns `undefined` after the app is restarted.
If we restart the app again the `getStringAsync` returns the expected value.
By restart I mean the app is closed by use with swipe up or is closed by to os and the user starts it again at a later time.

**To Reproduce**

1. Initialize the storage with the following
```javascript
this.storage = new MMKVStorage.Loader()
  .withServiceName('com.test.app1.TestV1')
  .withInstanceID('TestV1')
  .withEncryption()
  .initialize();
```
2. Call `this.storage.setStringAsync('token', tokenValue);` on login to store the secret token
3. Call `this.storage.getStringAsync('token');` on each app start, if the call returns a value it means the user is logged in
4. On some restarts at random calling `this.storage.getStringAsync('token');` returns undefined after the value is set.

**Expected behavior**
After calling `setStringAsync`, all the calls to `getStringAsync` should return the value set.

**Platform Information:**
 - OS: iOS
 - React Native Version: 0.65.1
 - Library Version v0.6.6

**Additional context**
We are also curious if other people experience the same issue?

**Comment by ammarahm-ed** [comment]
Are you calling `setStringAsync` immediately after `getStringAsync`? 

1. You need to add `await` for async functions.
2. If `setStringAsync` works, it should return true, otherwise it should returned `undefined`.
3. Check if mmkv is initialized properly by calling `console.log(global.setStringMMKV)`. It should return `[Function ...]` if everything is working as expected. 

Can you post here the exact stub of the function that you are using to call setStringAsync and getStringAsync?

**Comment by haikov** [comment]
> We are also curious if other people experience the same issue?

We're also experiencing something similar, in very rare cases, but so far haven't been able to reproduce it ourselves, only got it as a feedback.

We're using this library for persisting redux data, and every time affected users open the app, the storage seems to be clear and not persisted, which results in users logged out.

What we also noticed, reinstalling the app helps and so far is the only way for the user to recover. We started looking into it, but didn't find anything yet.

**Comment by andrei-tofan** [comment]
Hi @ammarahm-ed,

Thanks for the response.

> Are you calling `setStringAsync` immediately after `getStringAsync`?

No, we setStringAsync on login and getStringAsync when the app is started

> 1. You need to add `await` for async functions.

We do that

> 2. If `setStringAsync` works, it should return true, otherwise it should returned `undefined`.
> 3. Check if mmkv is initialized properly by calling `console.log(global.setStringMMKV)`. It should return `[Function ...]` if everything is working as expected.
> 
> Can you post here the exact stub of the function that you are using to call setStringAsync and getStringAsync?

Will create a demo app.

**Comment by ammarahm-ed** [comment]
We can fix this if we know what is causing the problem. Either storage is not initialized or it is not decrypted properly. Are you using an encrypted storage instance?

**Comment by andrei-tofan** [comment]
Many thanks!

> Are you using an encrypted storage instance?

Yes

**Comment by haikov** [comment]
> Are you using an encrypted storage instance?

Same for us. We're using it as a custom storage for redux persist as `new MMKVStorage.Loader().withEncryption().initialize()`

**Comment by ammarahm-ed** [comment]
@andrei-tofan @haikov It happens only to iOS users? How many users have been affected until now?

**Comment by haikov** [comment]
> It happens only to iOS users?

Yes, only iOS.

> How many users have been affected until now?

Hard to tell precisely, as we don't get any crash reports for it, but it's either coming from customer support, or from app reviews. So far, we have seen a few. And also once experienced ourselves, but didn't manage to debug anything because re-installing the app fixed the problem.

**Comment by andrei-tofan** [comment]
>  It happens only to iOS users?

Yes, only iOS.

> How many users have been affected until now?
Based on customer support I will estimate something around 5-10% / month, but I think some customers simply don't report the issue.

I've created a sample project with how we use react-native-mmkv-storage.
Let me know if we are doing something wrong
https://github.com/lineinspector/react-native-mmkv-storage-demo

Note that the issue can't be reproduced consistently.

**Comment by ammarahm-ed** [comment]
@andrei-tofan Have you been able to reproduce the issue at least once? Your implementation looks okay. No issue there.

I have an idea of what might be causing this. Will investigate.

**Comment by andrei-tofan** [comment]
Thx! @ammarahm-ed 

With the demo app not yet, but I'm running it on my phone and will update you once it happens.

**Comment by JoniVR** [comment]
I can confirm that we have faced this issue too. Over the past few months, I've had several users report "getting logged out randomly when launching the app". Also, one particular user said that when they restarted the app without logging in manually again, all was fine.

I've added more logging and from my testing, it does look like it's an issue related to reading persisted state. Couldn't figure out the reason at first either. At first I thought it was perhaps a redux-persist bug or refresh token failure bug, but after extensive logging that doesn't seem to be the issue.

I also can't reproduce it myself, but I've had about 8 reports on it out of about 1K users. 
Most of the reports are using the iPhone SE (think 2nd gen) but that could just be a coincidence. 
Also seems to be iOS only. We're using the encrypted storage with redux-persist. Hope this helps at least a bit.

**Comment by ammarahm-ed** [comment]
Hey @JoniVR thanks for adding this valuable information. It seems that sometimes mmkv is not registered. I will work a way to reproduce this then work on a solution. 8 times in 100 means if app is restarted about 100 times, maybe we can get 1 problematic launch.

**Comment by JoniVR** [comment]
Hi @ammarahm-ed!

Thanks for the quick reply. I'd also like to add to that that these specific users seem to be getting the issue more frequently than 8/100, perhaps even 2/10 times. Could also be more widespread but only these users reporting it.

Earlier reports were also saying it happend around when they received a push notification (though it also happens without one and even when you receive a push notification and just open the app regularly). 
Perhaps it's a configuration specific thing? (thinking low memory, low storage,...)

My guess was also around registration failing or even a timing issue where the storage isn't created in time for redux-persist (though not sure how plausible that is).

If there's anything I can do to help, let me know. We've set up a specific group of Testflight testers who are facing this issue and have asked them to report the issue each time it occurs (reporting in this case means sending logs). I'd be happy to test possible patches or add some logging if we need it.

**Comment by ammarahm-ed** [comment]
@JoniVR I will post here a list of things to Log. That will help in debugging.

**Comment by ammarahm-ed** [comment]
@JoniVR The very basic thing you should log:

```
console.log("storage initialized: ", global.getStringMMKV);
```
This should return `[Function ...]` always. The most probable reason for this being `undefined` could be MMKV not registering at runtime. Add this and next time a user reports. Check to logs to see if storage was initialized or not.

**Comment by JoniVR** [comment]
I've currently logged it like this in the latest test build (inside index.js): 

```js
if (typeof global.getStringMMKV !== 'function') {
  crashlytics().log(
    `global.getStringMMKV value: ${global.getStringMMKV}, see: https://github.com/ammarahm-ed/react-native-mmkv-storage/issues/195`,
  );
  crashlytics().recordError(
    new Error('typeof global.getStringMMKV is not a function'),
  );
} else {
  crashlytics().log(`MMKV storage initialized: ${global.getStringMMKV}`);
}
``` 
In case anyone else wants to use it. 

Now, let's hope a user runs into it again, should generate an automatic crashlytics report. If not, a manual report will confirm through logging that the function works as intended but the issue might be elsewhere.

**Comment by HT7-develop** [comment]
Currently a project i'm working on has been having this issue also ( but for the android version only, ios is working fine ).
As mentioned above by @ammarahm-ed  i tried logging : 
Which returns undefined (both have been used in App.js)
`console.log("storage initialized: ", global.getStringMMKV);`
and also returns undefined 
`await MMKV.setStringAsync("test", "some txt"); ,
const test = await MMKV.getStringAsync("isLoggedIn");
console.log('signin.js test: ', test);`
    
 Any additional info needed?
 
    Platform Information:
    OS: Android
    react-native: 0.63.4
    react-native-mmkv-storage: ^0.5.8,

**Comment by ammarahm-ed** [comment]
@HT7-develop You are getting this in the release build or while debugging?

**Comment by HT7-develop** [comment]
@ammarahm-ed thanks for the quick reply, I get it in the release build and also when debugging. 
(ios seems fine for both)

**Comment by ammarahm-ed** [comment]
@HT7-develop Does it happen always for you in debug? Is it always undefined?

**Comment by HT7-develop** [comment]
Storage has been working for the project for months without a problem, so both debug and release were working without problems.
About 2 months ago we noticed some issue with getting the user credentials, and when using the console log you provided (and any other we try to retrieve) is indeed always undefined.

**Comment by ammarahm-ed** [comment]
@HT7-develop How often does it happen while debugging and in release app? How many occurrences?

**Comment by HT7-develop** [comment]
It happens 100% of the time when logging in to the app for Android and the same for all the times I myself tried to retrieve something from the storage.

**Comment by ammarahm-ed** [comment]
@HT7-develop Are you sure you have configured everything correctly? You might want to recheck library configuration. You can also try to run the example app in the project which works normally. Also update the library to 0.6.10 and see if that helps?

**Comment by HT7-develop** [comment]
@ammarahm-ed , yep i'll try the example project and compare both projects for differences. and post an update here.

**Comment by ammarahm-ed** [comment]
We might have a potential fix to this issue. I will work on it asap.

**Comment by ammarahm-ed** [comment]
Hello, I have worked on a fix after the issue #207 was opened which brought me to a new and more predictable way to install JSI bindings. Basically the potential fix is taken from https://github.com/mrousavy/react-native-mmkv.

It's not released yet but you should test this with the master branch:

```
yarn add https://github.com/ammarahm-ed/react-native-mmkv-storage.git
```

After installation you will need to remove previous linking on Android.

`MainApplication.java`: 
Remove the commented out lines as below.
```java
package com.example;

import android.app.Application;
import android.content.Context;
import com.facebook.react.PackageList;
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactPackage;
//import com.facebook.react.bridge.JSIModulePackage; <---------------
import com.facebook.soloader.SoLoader;
import java.lang.reflect.InvocationTargetException;
import java.util.List;
//import com.ammarahmed.mmkv.RNMMKVJSIModulePackage; <---------------
public class MainApplication extends Application implements ReactApplication {

  private final ReactNativeHost mReactNativeHost =
      new ReactNativeHost(this) {
        @Override
        public boolean getUseDeveloperSupport() {
          return BuildConfig.DEBUG;
        }

        @Override
        protected List<ReactPackage> getPackages() {
          @SuppressWarnings("UnnecessaryLocalVariable")
          List<ReactPackage> packages = new PackageList(this).getPackages();
          // Packages that cannot be autolinked yet can be added manually here, for example:
          // packages.add(new MyReactNativePackage());
          return packages;
        }

        @Override
        protected String getJSMainModuleName() {
          return "index";
        }

      //    @Override <---------------
      //    protected JSIModulePackage getJSIModulePackage() { <---------------
      //        return new RNMMKVJSIModulePackage(); <---------------
       //   } <---------------
      };

  @Override
  public ReactNativeHost getReactNativeHost() {
    return mReactNativeHost;
  }

  ...
}

```
If you are using `reanimated@v2` then you should remove the modified `CustomRNMMKVJSIModulePackage` in `android/app/main/java/com/yourappid/` and remove any references to it in `MainApplication.java`.
```

**For iOS just run**:
```pod install
```
**Important Note**: Once you get the iOS build running. Drop a comment here and let me know that storage is working in debug and you are able to read/write data. I don't have a Mac with me at the moment so need a little help here.

**Comment by JoniVR** [comment]
Hi @ammarahm-ed, just installed on the iOS simulator without issues here, podfile does say `0.6.10` though, not sure if that's correct?

**Comment by ammarahm-ed** [comment]
@JoniVR I haven't bumped the version yet but this should be okay. You can confirm you have the latest version by checking if you can import `isLoaded` function from the library. 

`import {isLoaded} from "react-native-mmkv-storage"`

**Comment by JoniVR** [comment]
That seems to return `true`, so at least on debug and sim it seems to be working fine. Will push this change to beta testers. Anything specific you want me to log here? Last build didn't provide any results yet (only testing with 3 affected users).

**Comment by ammarahm-ed** [comment]
@JoniVR just log `isLoaded` function's result. It's basically doing the same thing you were doing above manually. `true` means that everything is working.

**Comment by ammarahm-ed** [comment]
Released **v0.6.11**

**Comment by haikov** [comment]
After upgrading to `0.6.11` we're able to build the app successfully and not able to reproduce the issue mentioned here (although we were unable to reproduce it previously either).

But, we started seeing some new runtime warning from inside MMKV: `-[UIApplication applicationState] must be used from main thread only`.

![image](https://user-images.githubusercontent.com/9068073/151793757-7c12f5cf-f42d-4c3a-b993-025ca10ed7f0.png)

**Comment by ammarahm-ed** [comment]
@haikov I will look into the warning.

**Comment by HT7-develop** [comment]
@ammarahm-ed , 
I also tested it (for android, as that is where storage the issue was for my project) and the issue seems resolved as of the latest update the storage is working again and saving/retrieving my users. 
In the debug build and release build, both are working fine.
Thanks for looking into it. really appreciate it.

OS: Android

**Comment by bombillazo** [comment]
> Hello, I have worked on a fix after the issue #207 was opened which brought me to a new and more predictable way to install JSI bindings. Basically the potential fix is taken from https://github.com/mrousavy/react-native-mmkv.
> 
> It's not released yet but you should test this with the master branch:
> 
> ```
> yarn add https://github.com/ammarahm-ed/react-native-mmkv-storage.git
> ```
> 
> After installation you will need to remove previous linking on Android.
> 
> `MainApplication.java`: Remove the commented out lines as below.
> 
> ```java
> package com.example;
> 
> import android.app.Application;
> import android.content.Context;
> import com.facebook.react.PackageList;
> import com.facebook.react.ReactApplication;
> import com.facebook.react.ReactInstanceManager;
> import com.facebook.react.ReactNativeHost;
> import com.facebook.react.ReactPackage;
> //import com.facebook.react.bridge.JSIModulePackage; <---------------
> import com.facebook.soloader.SoLoader;
> import java.lang.reflect.InvocationTargetException;
> import java.util.List;
> //import com.ammarahmed.mmkv.RNMMKVJSIModulePackage; <---------------
> public class MainApplication extends Application implements ReactApplication {
> 
>   private final ReactNativeHost mReactNativeHost =
>       new ReactNativeHost(this) {
>         @Override
>         public boolean getUseDeveloperSupport() {
>           return BuildConfig.DEBUG;
>         }
> 
>         @Override
>         protected List<ReactPackage> getPackages() {
>           @SuppressWarnings("UnnecessaryLocalVariable")
>           List<ReactPackage> packages = new PackageList(this).getPackages();
>           // Packages that cannot be autolinked yet can be added manually here, for example:
>           // packages.add(new MyReactNativePackage());
>           return packages;
>         }
> 
>         @Override
>         protected String getJSMainModuleName() {
>           return "index";
>         }
> 
>       //    @Override <---------------
>       //    protected JSIModulePackage getJSIModulePackage() { <---------------
>       //        return new RNMMKVJSIModulePackage(); <---------------
>        //   } <---------------
>       };
> 
>   @Override
>   public ReactNativeHost getReactNativeHost() {
>     return mReactNativeHost;
>   }
> 
>   ...
> }
> ```
> 
> If you are using `reanimated@v2` then you should remove the modified `CustomRNMMKVJSIModulePackage` in `android/app/main/java/com/yourappid/` and remove any references to it in `MainApplication.java`.
> 
> ```
> 
> **For iOS just run**:
> ```pod install
> ```
> 
> **Important Note**: Once you get the iOS build running. Drop a comment here and let me know that storage is working in debug and you are able to read/write data. I don't have a Mac with me at the moment so need a little help here.

If we were using a previous version, do we need to follow these steps to upgrade to >= 6.11?

To confirm, we no longer need this file? 
![image](https://user-images.githubusercontent.com/44002321/152115539-919adb86-9e2d-45a0-85ad-6f1c7228d946.png)

**Comment by ammarahm-ed** [comment]
Yes you don't need this. However if you don't remove it, everything should still work without changing anything. @bombillazo

**Comment by bombillazo** [comment]
Ive updated my MainApplication.java file but this problem still persist for me, I see the store working while the app is running but every time I run my loading logic when the app opens, the stores seem o be empty or something.

**EDIT**

So printing `isLoaded` returns false. I installed it and ran `npx mmkv-link` and the build was successful. I checked the files manually as well. What could be causing `isLoaded` to return false? @ammarahm-ed 

"react-native": "0.67.1",

**Comment by ammarahm-ed** [comment]
@bombillazo 

run `./gradlew clean` on `/android` folder. 
Delete `node_modules`. 
Delete app from your phone. 
Then run `yarn install` or `npm install`. 
Finally install the app again and everything should be good.

Also you don't need to run `npx mmkv link` on RN 0.67 and above.

Also make sure to run `pod install` on iOS.


Did this happen on iOS or Android?

**Comment by bombillazo** [comment]
Its on Android, I did not reinstall the app, let me try that, thanks.

**Comment by bombillazo** [comment]
Still set to `false` :/ 

On older versions, it did return something when I called { global.set...} when I had `import com.ammarahmed.mmkv.RNMMKVJSIModulePackage;` in my MainApplication.java file.

Is there any way to check the native files manually to verify if it installed correctly?

**Comment by ammarahm-ed** [comment]
@bombillazo Try to log `NativeModules.MMKVNative.install`:

```
import {NativeModules} from "react-native";
console.log(NativeModules.MMKVNative.install);
```

**Comment by bombillazo** [comment]
> @bombillazo Try to log `NativeModules.MMKVNative.install`:
> 
> ```
> import {NativeModules} from "react-native";
> console.log(NativeModules.MMKVNative.install);
> ```

I get this back
```
 function nonPromiseMethodWrapper() { [bytecode] }
```

**Comment by ammarahm-ed** [comment]
@bombillazo Seems like everything is updated correctly. Try this:

```js
import {init,isLoaded} from "react-native-mmkv-storage";

console.log(init(),isLoaded(), global.setStringMMKV)
```

**Comment by bombillazo** [comment]
> @bombillazo Seems like everything is updated correctly. Try this:
> 
> ```js
> import {init,isLoaded} from "react-native-mmkv-storage";
> 
> console.log(init(),isLoaded(), global.setStringMMKV}
> ```

wow, ok after running it the first time, now it is returning true! Even when I close the app or clear the app data.

**Comment by ammarahm-ed** [comment]
@bombillazo Seems like your metro cache is old causing a problem when you upgrade loading old bundled js files. Restart metro bundler with cache reset:

```
yarn start --reset-cache
```
If you don't know how that works. Just remove `node_modules` folder and run `npm/yarn install`

And now everything should work as expected.

**Comment by bombillazo** [comment]
Its weird, I always run metro with --reset-cache for that same reason, I don't trust the caches. But glad it works now! Thank you for the assistance!

**Comment by ammarahm-ed** [comment]
Most welcome. Do let me know if you face any other issues

**Comment by JoniVR** [comment]
Hi @ammarahm-ed 

I've just had 2 reports in the same day from one of our testers telling me that the issue still seems to exist. Not sure if it's related to my specific codebase though.

Specifics:
* No values are read from storage (default reducer values are used).
* Killing the app and restarting it restores data correctly.
* The `isLoaded()` function returns `true` in `index.js`
* Platform is iOS

Below is one of the logfiles that demonstrate the scenario:
[be.frontforce.emergency.mobile_issue_f26bcd4da6f466433e7f065611c4e778_error_session_81fe6016199e44fc9abb239326ff06f8_DNE_0_v2.log](https://github.com/ammarahm-ed/react-native-mmkv-storage/files/8033330/be.frontforce.emergency.mobile_issue_f26bcd4da6f466433e7f065611c4e778_error_session_81fe6016199e44fc9abb239326ff06f8_DNE_0_v2.log)


Could it perhaps still be the `getStringAsync` function? Should I log this?
Any other values I can log to identify other potential issues?

**Comment by ammarahm-ed** [comment]
@JoniVR Is your instance encrypted or not encrypted?

**Comment by JoniVR** [comment]
Encrypted, setup code:

```ts
const encryptedStorage = new MMKVStorage.Loader().withEncryption().initialize();
```

```ts
const authPersistConfig: PersistConfig<AuthState> = {
  key: 'auth',
  version: 2,
  storage: encryptedStorage,
  stateReconciler: autoMergeLevel2,
  writeFailHandler,
  timeout: 12000,
};
```

**Comment by ammarahm-ed** [comment]
This looks ok. Nothing unusual. Also the initialization seems ok too. 

The `getStringAsync` function returns `undefined` or `null` when values from storage do not load? 

To debug more, do the following:

```
// Log which instances are loaded currently or if none are loaded.
console.log(MMKVStorage.getCurrentMMKVInstanceIDs());

// Check our id store is working
console.log(encryptedStorage.getAllMMKVInstanceIDs());

// Manually extract key from keychain: should always return true;
let key = encryptedStorage.getKey();
if (key.alias) {
   console.log(global.secureKeyExists(key.alias));
}

// Check what is returned from getString function. This value does not need to be written in storage. If it returns null, means storage is ok. If it returns undefined. It means storage instance is not loaded.
console.log(encryptedStorage.getString("unknown_value")); 
```
@JoniVR

**Comment by JoniVR** [comment]
Thanks @ammarahm-ed, will add these!

**Comment by ammarahm-ed** [comment]
@JoniVR Make sure to test them in debug mode first.

**Comment by JoniVR** [comment]
@ammarahm-ed I did, they seem to all be printing the expected values:
```
 DEBUG  2022-02-09T17:52:30 | index.js | MMKV storage initialized: true
 DEBUG  2022-02-09T17:52:30 | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":true}
 DEBUG  2022-02-09T17:52:30 | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 DEBUG  2022-02-09T17:52:30 | MMKV:global.secureKeyExists | true
 DEBUG  2022-02-09T17:52:30 | MMKV:encryptedStorage.getString | null
``` 

I've set up automated crashlytics reporting for unexpected values again, let bug hunting season recommence! 😄

**Comment by ammarahm-ed** [comment]
Hey @JoniVR Did you get any reports on the issue yet?

**Comment by JoniVR** [comment]
@ammarahm-ed Nothing as of yet (no manual reports either), though I did ask the user facing the issue to reinstall the app. Could this perhaps somehow still be related to old state being incorrectly created? (highly doubt it though). Anyways, am monitoring on a daily basis and will keep you posted in case it does occur.

**Comment by JoniVR** [comment]
Hi @ammarahm-ed! 

Just had a report where `getString` didn't return `null` but `undefined`. See logs:

```
# Crashlytics - Custom logs
# Application: be.frontforce.emergency.mobile
# Platform: apple
# Version: 1.2.1 (88)
# Date: Mon Feb 21 2022 00:14:47 GMT+0100

 0 | Mon Feb 21 2022 00:13:19 GMT+0100 | DEBUG | index.js | MMKV storage initialized: true
 1 | Mon Feb 21 2022 00:13:19 GMT+0100 | DEBUG | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":false}
 2 | Mon Feb 21 2022 00:13:19 GMT+0100 | DEBUG | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 3 | Mon Feb 21 2022 00:13:19 GMT+0100 | DEBUG | MMKV:global.secureKeyExists | true
 4 | Mon Feb 21 2022 00:13:19 GMT+0100 | DEBUG | MMKV:encryptedStorage.getString | undefined
 5 | Mon Feb 21 2022 00:13:19 GMT+0100 | ERROR | MMKV:encryptedStorage.getString | Expected null is not null
 6 | Mon Feb 21 2022 00:13:20 GMT+0100 | DEBUG | persistErrorHandlerMiddleware | Rehydrating auth, token: null
 7 | Mon Feb 21 2022 00:13:20 GMT+0100 | DEBUG | App | Notification permission status: 1
 8 | Mon Feb 21 2022 00:13:20 GMT+0100 | DEBUG | setMessagingToken | OS: ios, Token: cr5hJb...5x9KLv
 9 | Mon Feb 21 2022 00:13:21 GMT+0100 | DEBUG | httpInterceptor:responseHandler | {
"requestUrl": "/endpoints.emergency.json",
"responseCode": 200,
"bearer": null
}
```

**Comment by ammarahm-ed** [comment]
Thanks. A fix is already on master branch for this!

**Comment by JoniVR** [comment]
You're amazing! I'll just create a new beta using a version pointed at master for now then?

**Comment by ammarahm-ed** [comment]
@JoniVR You can use the master branch however please note that there are numerous changes in upcoming v0.7.0 of library as mentioned in #213. However none are breaking so it should be good. If you face any issues while using the master branch, open a separate issue for it.

**Comment by JoniVR** [comment]
@ammarahm-ed Just had another report with about the same logs in the new build using master:


```
# Crashlytics - Custom logs
# Application: be.frontforce.emergency.mobile
# Platform: apple
# Version: 1.2.1 (90)
# Issue: 853db40e9631ddf8bdb291375ee5a1fb
# Session: 15dbab09bcc74541a9c72e23a02aed88_DNE_0_v2
# Date: Mon Feb 21 2022 16:34:46 GMT+0100

0 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | index.js | MMKV storage initialized: true
1 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":false}
2 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
3 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | MMKV:global.secureKeyExists | true
4 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | MMKV:encryptedStorage.getString | undefined
5 | Mon Feb 21 2022 16:34:46 GMT+0100 | ERROR | MMKV:encryptedStorage.getString | Expected null is not null
6 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | persistErrorHandlerMiddleware | Rehydrating auth, token: null
7 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | App | Notification permission status: 1
8 | Mon Feb 21 2022 16:34:46 GMT+0100 | DEBUG | setMessagingToken | OS: ios, Token: drg0AG...OKAwpu
```

Also note how `getCurrentMMKVInstanceIDs`  returns `{"default":false}`

<img width="1211" alt="image" src="https://user-images.githubusercontent.com/7591717/155111844-01d1daf7-f8c1-4efb-b6c2-dcbc4634e215.png">

<img width="856" alt="image" src="https://user-images.githubusercontent.com/7591717/155111904-7cc3616a-aa92-4603-b09e-f1d05b01250c.png">

I had messed something up with automatic reporting in previous builds, that's why it only showed up after one of the users manually reported it. This one's automatic.

**Comment by ammarahm-ed** [comment]
@JoniVR I just reviewed the changes I made and found a loophole which made the fix not work. I have fixed it now.

Install from master again:
```
npm install https://github.com/ammarahm-ed/react-native-mmkv-storage.git
```
Run `pod install`

**Comment by JoniVR** [comment]
Hi @ammarahm-ed, got another report of the same issue in the newest build:

```
# Crashlytics - Custom logs
# Application: be.frontforce.emergency.mobile
# Platform: apple
# Version: 1.2.1 (91)
# Issue: 853db40e9631ddf8bdb291375ee5a1fb
# Session: 25047b061e984398804e167507878dfe_DNE_0_v2
# Date: Tue Feb 22 2022 19:16:59 GMT+0100

 0 | Tue Feb 22 2022 19:16:58 GMT+0100 | DEBUG | index.js | MMKV storage initialized: true
 1 | Tue Feb 22 2022 19:16:58 GMT+0100 | DEBUG | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":false}
 2 | Tue Feb 22 2022 19:16:58 GMT+0100 | DEBUG | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 3 | Tue Feb 22 2022 19:16:58 GMT+0100 | DEBUG | MMKV:global.secureKeyExists | true
 4 | Tue Feb 22 2022 19:16:58 GMT+0100 | DEBUG | MMKV:encryptedStorage.getString | undefined
 5 | Tue Feb 22 2022 19:16:58 GMT+0100 | ERROR | MMKV:encryptedStorage.getString | Expected null is not null
 6 | Tue Feb 22 2022 19:16:59 GMT+0100 | DEBUG | persistErrorHandlerMiddleware | Rehydrating auth, token: null
 7 | Tue Feb 22 2022 19:16:59 GMT+0100 | DEBUG | App | Notification permission status: 1
 8 | Tue Feb 22 2022 19:16:59 GMT+0100 | DEBUG | setMessagingToken | OS: ios, Token: fSK95L...x9VUfO
 9 | Tue Feb 22 2022 19:16:59 GMT+0100 | DEBUG | httpInterceptor:responseHandler | {
"requestUrl": "/endpoints.emergency.json",
"responseCode": 200,
"bearer": null
}
```
Different user again :/

**Comment by ammarahm-ed** [comment]
@JoniVR If I add `console.log` in the library code, will you be able to log them to your backend as above automatically?

**Comment by JoniVR** [comment]
@ammarahm-ed I doubt it unless I start intercepting console logs (I'm using Firebase Crashlytics) but that'll probably clutter the logs. 

However I do think I can log  using Crashlytics at the native layer too (or JS layer, wherever you want), so perhaps if you tell me where and what you want to log I can patch package it in my project?

**Comment by ammarahm-ed** [comment]
@JoniVR I have added logging in the library in a seperate branch:

Install as follows:

```
npm install https://github.com/ammarahm-ed/react-native-mmkv-storage#bugfix
```
Usage:

```jsx
const encryptedStorage = new MMKVStorage.Loader().withEncryption().initialize((logs) => {
         console.log(logs); // Will print a single string with all the logs at the end of an operation. You can replace this with crashlytics.
});
```
Make sure everything logs correctly in debug.

**Comment by JoniVR** [comment]
Output in debug:
```
LOG  Initialize: started
status: storage instance already exists
status: alias:true, trying to init existing instance with old key
status: key: true alias:true
status: key found for storage
status: Encrypted instance loaded true, verifying
status: Encrypted instance loaded and verified
 LOG  storage instance default is loaded?: true
```

Published a new release and will keep you updated

**Comment by ammarahm-ed** [comment]
@JoniVR some logs are missing. Have you shared all the logs or not? There is also logging on get/set a value.

**Comment by JoniVR** [comment]
That is purely the logging I get from using the logs parameter. 
This is on app launch (remember that I'm using redux-persist). 

Only my auth store is using the encrypted storage.
I left out any other logging that isn't from that parameter.

**Comment by ammarahm-ed** [comment]
@JoniVR It should also log getting the token value from storage. If it's undefined, null etc.

```js
opts.logs.push(
    `result for ${args[0]} is: undefined?${result === undefined} null?${
      result === null
    }: current instance status: ${currentInstancesStatus[id]}`
  );
```
It's something like this as implemented. The callback might be called multiple times.

**Comment by JoniVR** [comment]
This is all I get in debug on my device unfortunately. I did do a fresh npm & pod install just to be sure.
```
 DEBUG  2022-02-25T09:56:09 | MMKV:logs | Initialize: started
status: storage instance already exists
status: alias:true, trying to init existing instance with old key
status: key: true alias:true
status: key found for storage
status: Encrypted instance loaded true, verifying
status: Encrypted instance loaded and verified
 DEBUG  2022-02-25T09:56:09 | MMKV:logs | storage instance default is loaded?: true
 DEBUG  2022-02-25T09:56:09 | index.js | MMKV storage initialized: true
 DEBUG  2022-02-25T09:56:09 | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":true}
 DEBUG  2022-02-25T09:56:09 | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 DEBUG  2022-02-25T09:56:09 | MMKV:global.secureKeyExists | true
 DEBUG  2022-02-25T09:56:09 | MMKV:encryptedStorage.getString | null
 LOG  2022-02-25T09:56:09 | Reactotron debugger configured.
 DEBUG  2022-02-25T09:56:09 | persistErrorHandlerMiddleware | Rehydrating auth, token: XXXXXX...XXXXXX
 LOG  Running "ffemergency" with {"rootTag":1,"initialProps":{}}
 DEBUG  2022-02-25T09:56:09 | AuthService:refreshAuthToken | Attempting token refresh with: XXXXXX...XXXXXX
 DEBUG  2022-02-25T09:56:10 | httpInterceptor:responseHandler | {
  "requestUrl": "/token",
  "responseCode": 200,
  "bearer": "XXXXXX...XXXXXX"
}
 DEBUG  2022-02-25T09:56:10 | TokenRefreshHandler:refreshAuthToken | Refresh token request successful. "bearer": "XXXXXX...-XXXXXX", refresh: XXXXXX...XXXXXX
 DEBUG  2022-02-25T09:56:10 | authActions | tokenRefresh: Bearer: "XXXXXX...-XXXXXX", refresh: XXXXXX...XXXXXX
 DEBUG  2022-02-25T09:56:10 | TokenRefreshHandler:retryFailedRequests | Retrying failed requests
 DEBUG  2022-02-25T09:56:10 | TokenRefreshHandler:clearRetryQueue | Clearing request retry queue
 DEBUG  2022-02-25T09:56:10 | App | Notification permission status: 1
 DEBUG  2022-02-25T09:56:10 | setMessagingToken | OS: ios, Token: fDuhqXXXXXX...XXXXXX
```

**Comment by JoniVR** [comment]
I see the issue now, the yarn.lock file hash points to the second last commit:
<img width="809" alt="image" src="https://user-images.githubusercontent.com/7591717/155686964-794db57a-4456-4173-b9f5-0aa92d85cfc5.png">

instead of 3b1d208d0770cfc9223710a90943d0a1ae6ce4b6

Will make sure to get latest commit, check logs for more details and submit a new test build, my bad

**Edit: logs are still lacking the specific lines you mentioned, no clue why.**

**Comment by JoniVR** [comment]
I double checked inside the node modules, your specified line is present in the installed module but the log is not appearing.

Did a full removal of node_modules and metro start with clean cache several times now, so it should be fine on that end.

**Comment by ammarahm-ed** [comment]
@JoniVR Here is how the example app is logging everything: 

The instance being logged:
```jsx
const storage = new MMKVStorage.Loader()
  .withEncryption()
  .initialize(logs => console.log(logs));
```

Logs:
```
Initialize: started
status: storage instance already exists
status: storage instance is not encrypted, initialize without encryption
status: ready: true
status: instance verified
 LOG  storage instance default is loaded?: true
 LOG  Running "example" with {"rootTag":11}
 LOG  status: fetch value for user
result for user is: undefined?false null?false: current instance status: true
 LOG  status: fetch value for stringIndex
result for stringIndex is: undefined?false null?false: current instance status: true
 LOG  status: fetch value for user
result for user is: undefined?false null?false: current instance status: true
 LOG  status: fetch value for user
result for user is: undefined?false null?false: current instance status: true
 LOG  status: fetch value for stringIndex
result for stringIndex is: undefined?false null?false: current instance status: true
 LOG  status: fetch value for age
result for age is: undefined?false null?false: current instance status: true
 LOG  status: fetch value for age
result for age is: undefined?false null?false: current instance status: true
```

**Comment by JoniVR** [comment]
@ammarahm-ed I just swapped out my entire App.tsx file with the one in your example code and then it works fine, yet when running with redux persist in the exact same setup I'm not getting these logs.

Though I only use mmkv for the auth store (still have to write a migration from AsyncStorage for the other stores).

**Comment by JoniVR** [comment]
@ammarahm-ed Got a new report (2 actually but they're exactly the same):

```log
 0 | Sun Feb 27 2022 21:02:37 GMT+0100 | DEBUG | MMKV:logs | Initialize: started
status: storage instance already exists
status: alias:true, trying to init existing instance with old key
status: key: false alias:true
Error: key does not exist in keychain
 1 | Sun Feb 27 2022 21:02:37 GMT+0100 | DEBUG | MMKV:logs | storage instance default is loaded?: false
 2 | Sun Feb 27 2022 21:02:37 GMT+0100 | DEBUG | index.js | MMKV storage initialized: true
 3 | Sun Feb 27 2022 21:02:37 GMT+0100 | DEBUG | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":false}
 4 | Sun Feb 27 2022 21:02:37 GMT+0100 | DEBUG | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 5 | Sun Feb 27 2022 21:02:37 GMT+0100 | DEBUG | MMKV:global.secureKeyExists | true
 6 | Sun Feb 27 2022 21:02:37 GMT+0100 | DEBUG | MMKV:encryptedStorage.getString | undefined
 7 | Sun Feb 27 2022 21:02:37 GMT+0100 | ERROR | MMKV:encryptedStorage.getString | Expected null is not null
 8 | Sun Feb 27 2022 21:02:38 GMT+0100 | DEBUG | persistErrorHandlerMiddleware | Rehydrating auth, token: null
 9 | Sun Feb 27 2022 21:02:38 GMT+0100 | DEBUG | App | Notification permission status: 1
10 | Sun Feb 27 2022 21:02:38 GMT+0100 | DEBUG | setMessagingToken | OS: ios, Token: eGBZXb...8HXx-y
```

Still doesn't include all logging you wanted but I do think it offers some extra insight

**Comment by ammarahm-ed** [comment]
@JoniVR I think we have gotten closer to this bug. It seems that for some reason `key does not exist in keychain` at the very beginning of app launch sometimes but only a few milliseconds later `MMKV:global.secureKeyExists | true` is returned, meaning the key is there but was not read initially for some reason.

**Comment by JoniVR** [comment]
@ammarahm-ed Correct, I can also tell you with a fair amount of certainty that it only affects iOS users. I have about 100 Android beta testers and around 15 iOS TestFlight testers, all reports are coming from this small group of iOS testers.

**Comment by ammarahm-ed** [comment]
@JoniVR I have pushed some changes to the #bugfix branch to add some more logging. Pull and release a new version.

Also try to see why you aren't getting logs on fetching values? Its strange.

One more log to add:

```
console.log(global.getSecureKey(key.alias));
```
Add this in the `key.alias` if block.

**Comment by JoniVR** [comment]
New report with changes you requested:
```
 0 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | MMKV:logs | Initialize: started
status: storage instance already exists
status: alias:true, trying to init existing instance with old key
status: keyExists:true: key: false alias:true
Error: key does not exist in keychain: keyExists:true: key: false alias:true
 1 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | MMKV:logs | storage instance default is loaded?: false
 2 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | index.js | MMKV storage initialized: true
 3 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":false}
 4 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 5 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | MMKV:global.secureKeyExists | true
 6 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | MMKV:global.getSecureKey |
 7 | Fri Mar 04 2022 07:49:19 GMT+0100 | DEBUG | MMKV:encryptedStorage.getString | undefined
 8 | Fri Mar 04 2022 07:49:19 GMT+0100 | ERROR | MMKV:encryptedStorage.getString | Expected null is not null
 9 | Fri Mar 04 2022 07:49:20 GMT+0100 | DEBUG | persistErrorHandlerMiddleware | Rehydrating auth, token: null
10 | Fri Mar 04 2022 07:49:20 GMT+0100 | DEBUG | App | Notification permission status: 1
11 | Fri Mar 04 2022 07:49:20 GMT+0100 | DEBUG | setMessagingToken | OS: ios, Token: cw04OI...RMH0zz
```

From what I can tell, the value is completely empty. Not even undefined?

> Also try to see why you aren't getting logs on fetching values? Its strange.

I did check a lot of stuff and tried a lot of things, like I said, I even replaced my entire `App.tsx` with the one from your example and then it worked, so the code is there, it just isn't getting called for some reason. Will look into it more when I get more time on the project, have currently shifted priorities a bit to another project but will continue following up on this.

**Comment by ammarahm-ed** [comment]
@JoniVR So after some investigation it seems that the issue is not with this library but it's an issue on iOS 15+ devices. 

Basically on iOS 15 a prewarm feature was added, which is causing the keychain to not return values sometimes on app launch or the values being nil etc. The logs above tell a similar story.

As you can see, the logs say that the key exists but it is null which at first is strange but now that i have researched on google. It seems many people are facing this. I will investigate more.

https://www.google.com/amp/s/researchsnipers.com/ios-15-gets-weird-login-problems/amp/

**Comment by JoniVR** [comment]
Interesting find! That might explain it all, more in this Twitter post:

https://twitter.com/steipete/status/1466013492180312068?s=21

**Comment by JoniVR** [comment]
Potential fix?

https://github.com/evgenyneu/keychain-swift/issues/78#issuecomment-549045569

Also linked:

https://github.com/mCodex/react-native-sensitive-info/pull/183

**Comment by ammarahm-ed** [comment]
@JoniVR another potential fix https://theiosdude.medium.com/ios-confessions-episode-2-ios15-prewarming-logouts-and-our-answer-to-the-keychain-not-available-43fcca4420c5

But before that, we need to update the logging once to get the actual error from the native iOS code to JS so we know the exact error.

**Comment by JoniVR** [comment]
Best thing to log would probably be `UIApplication.shared.isProtectedDataAvailable` inside `didFinishLaunchingWithOptions` would be my guess?

**Comment by JoniVR** [comment]
@ammarahm-ed I logged the `isProtectedDataAvailable` value and passed it with `initialProps` from the native layer. The generated report does indeed indicate that this value is `false` when the bug occurs.

Log:
```log
 0 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | MMKV:logs | Initialize: started
status: storage instance already exists
status: alias:true, trying to init existing instance with old key
status: keyExists:true: key: false alias:true
Error: key does not exist in keychain: keyExists:true: key: false alias:true
 1 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | MMKV:logs | storage instance default is loaded?: false
 2 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | index.js | MMKV storage initialized: true
 3 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":false}
 4 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 5 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | MMKV:global.secureKeyExists | true
 6 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | MMKV:global.getSecureKey |
 7 | Fri Mar 11 2022 13:15:57 GMT+0100 | DEBUG | MMKV:encryptedStorage.getString | undefined
 8 | Fri Mar 11 2022 13:15:57 GMT+0100 | ERROR | MMKV:encryptedStorage.getString | Expected null is not null
 9 | Fri Mar 11 2022 13:15:58 GMT+0100 | DEBUG | persistErrorHandlerMiddleware | Rehydrating auth, token: null
10 | Fri Mar 11 2022 13:15:58 GMT+0100 | DEBUG | App:onBeforeLift | isProtectedDataAvailable: false
11 | Fri Mar 11 2022 13:15:58 GMT+0100 | ERROR | App:onBeforeLift | isProtectedDataAvailable: false
12 | Fri Mar 11 2022 13:15:58 GMT+0100 | DEBUG | App | Notification permission status: 1
13 | Fri Mar 11 2022 13:15:58 GMT+0100 | DEBUG | setMessagingToken | OS: ios, Token: drg0AG...OKAwpu
14 | Fri Mar 11 2022 13:15:58 GMT+0100 | DEBUG | httpInterceptor:responseHandler | {
"requestUrl": "/endpoints.emergency.json",
"responseCode": 200,
"bearer": null
}
```

**Comment by ammarahm-ed** [comment]
@JoniVR I think this means that persist store is rehydrated too soon in app lifecycle causing a race condition on some devices randomly. Is there a way to delay rehydrating the persist store?

**Comment by JoniVR** [comment]
@ammarahm-ed If I understand its inner workings correctly, to support redux-persist you simply provide `getItem` and `setItem` API's right? The storage instance is then passed to redux persist. 

How does redux persist know when the storage engine is ready? Is it just passing a reference and not caring about the state of the storage engine?

I think the approach with listening for `isProtectedDataAvailable` is a good solution. Can that value perhaps be tracked somewhere internally in mmkv and then get checked when `getItem` is called and delay the function there? I'm guessing that would be an approach since these functions can be async? Extra benefit is that it will also prevent the issue from happening outside of redux-persist cases.

**Comment by ammarahm-ed** [comment]
@JoniVR The api is synchronous so storage is ready when called usually. I will add a workaround for the keychain issue as you have suggested.

**Comment by JoniVR** [comment]
@ammarahm-ed Where will you be implementing the fix? In the `getItem` function or rather just when initialising the storage? I looked at the code a bit but couldn't really figure out how to delay storage initialization itself.

**Comment by JoniVR** [comment]
Not sure if this is the correct way, but in our latest test version I've added a quick workaround (or at least I hope so):

```patch
diff --git a/node_modules/react-native-mmkv-storage/ios/SecureStorage.m b/node_modules/react-native-mmkv-storage/ios/SecureStorage.m
index eacd030..337bee2 100644
--- a/node_modules/react-native-mmkv-storage/ios/SecureStorage.m
+++ b/node_modules/react-native-mmkv-storage/ios/SecureStorage.m
@@ -5,7 +5,7 @@
 #endif
 
 #import "SecureStorage.h"
-
+#import <UIKit/UIKit.h>
 
 @implementation SecureStorage : NSObject
 
@@ -43,6 +43,14 @@ - (NSString *) getSecureKey:(NSString *)key
     
     @try {
         [self handleAppUninstallation];
+        dispatch_sync(dispatch_get_main_queue(), ^{
+            int readAttempts = 0;
+            // See: https://github.com/ammarahm-ed/react-native-mmkv-storage/issues/195
+            while (![[UIApplication sharedApplication] isProtectedDataAvailable] && readAttempts++ < 10) {
+                // sleep 250ms before another attempt
+                usleep(250000);
+            }
+        });
         NSString *value = [self searchKeychainCopyMatching:key];
         if (value == nil) {
             NSString* errorMessage = @"key does not present";
```

**Comment by ammarahm-ed** [comment]
@JoniVR `dispatch_async` will not work since it is `async` and the native call is synchronous so while you are trying to delay the return value, it will not be delayed actually. However doing the same in `setItem` or `getItem` in JS will work since they are `async`. 

In native objc++ code, maybe we can use `std::thread thread_obj(lamda)` and then call `thread_obj.join()` to make it wait for the result before returning value.

I have been a bit busy. I plan to add a fix on Sunday which should work properly.

**Comment by JoniVR** [comment]
@ammarahm-ed That's why I used `dispatch_sync` instead, I figured blocking the main thread should also block JS execution? Or is this a wrong assumption? (I don't care about blocking main thread because my storage is initialized at app launch anyway)

I tested it with a 20s sleep and it seemed to block app loading. Only thing I'm unsure of is if it effectively blocks JS and also doesn't block the actual `isProtectedDataAvailable` from updating too.

**Comment by ammarahm-ed** [comment]
@JoniVR I think I misread your code a bit. `dispatch_sync` should work too similar to `thread.join()`. Since you have added a workaround. let's wait a few days and see if you get any reports of the issue. I hope there won't be any. We can merge the above logic in the library.

I also think that `250` is a very big value, it should check every 10-50ms instead with iterations ranging 10-50. So response is instant.

**Comment by JoniVR** [comment]
@ammarahm-ed No problem, it's easily misread and `dispatch_sync` is not used that often compared to the async variant. I deployed the possible fix about 48 hours ago, nothing reported so far.. let's hope it remains that way, usually takes a few days before they start popping up 🤞🏻

> I also think that 250 is a very big value, it should check every 10-50ms instead with iterations ranging 10-50. So response is instant.

I agree with you, will change it after confirming it solves the issue (or just whenever the fix is implemented in the library) 😀

**Comment by JoniVR** [comment]
Update: So far so good, no reports yet (neither manual nor automatic) 😄

**Comment by jaltin** [comment]
@JoniVR, great to hear that your fix seems to work well so far. @ammarahm-ed do you think this is ready to bake into a new release soon?

**Comment by ammarahm-ed** [comment]
@jaltin I am working on it.

**Comment by ammarahm-ed** [comment]
@JoniVR I think the workaround needs to be added for both get/set/remove methods in `SecureStorage.h`.

**Comment by JoniVR** [comment]
@ammarahm-ed Probably yes. My workaround works for my specific case due to the fact that I just need to delay app launch and wait for the storage engine (storage engine is used in redux-persist and splash screen only disappears when redux-persist is done) This is why I labeled it as a workaround and didn't submit a PR yet.

For other users, this might not be feasible as the engine might be set up at slightly a later point, potentially freezing the entire UI thread while the app is already open.

I just wanted to try out if this simple approach worked, because we could also observe the notification, but that would complicate things since you'd have to make your code await the notification. Polling seems easier in this case, but you probably can't block the entire UI thread for it. So since get/set/remove are already async anyway, doing this check there seems like a better decision?

**Comment by ammarahm-ed** [comment]
@JoniVR The best way would be to listen to notification but since **initialization is not async** and many people depend on the synchronous api I don't see another way of doing this. I think that app should not launch before protected data is available because most applications will depend on that data. 

If the engine is setup at a later point, the workaround will have no effect as protected data would already be available. get/set/remove are not async either which is why a listener would not work.

Btw are you seeing any unnecessary delays in app launch after you made the change in release?

**Comment by JoniVR** [comment]
Also interesting, I just got a new report on the patched build, check out the logging:

```
0 | Mon Apr 04 2022 22:43:06 GMT+0200 | DEBUG | MMKV:logs | Initialize: started
status: storage instance already exists
status: alias:true, trying to init existing instance with old key
status: keyExists:true: key: false alias:true
Error: key does not exist in keychain: keyExists:true: key: false alias:true
 1 | Mon Apr 04 2022 22:43:12 GMT+0200 | DEBUG | MMKV:logs | storage instance default is loaded?: false
 2 | Mon Apr 04 2022 22:43:12 GMT+0200 | DEBUG | index.js | MMKV storage initialized: true
 3 | Mon Apr 04 2022 22:43:15 GMT+0200 | DEBUG | MMKV:MMKVStorage.getCurrentMMKVInstanceIDs | {"default":false}
 4 | Mon Apr 04 2022 22:43:15 GMT+0200 | DEBUG | MMKV:encryptedStorage.getAllMMKVInstanceIDs | default
 5 | Mon Apr 04 2022 22:43:15 GMT+0200 | DEBUG | MMKV:global.secureKeyExists | true
 6 | Mon Apr 04 2022 22:43:15 GMT+0200 | DEBUG | MMKV:global.getSecureKey |
 7 | Mon Apr 04 2022 22:43:28 GMT+0200 | DEBUG | MMKV:encryptedStorage.getString | undefined
 8 | Mon Apr 04 2022 22:43:28 GMT+0200 | ERROR | MMKV:encryptedStorage.getString | Expected null is not null
 9 | Mon Apr 04 2022 22:53:25 GMT+0200 | DEBUG | persistErrorHandlerMiddleware | Rehydrating auth, token: eyJhbG...ciYRw8
10 | Mon Apr 04 2022 22:53:25 GMT+0200 | DEBUG | App:onBeforeLift | isProtectedDataAvailable: false
11 | Mon Apr 04 2022 22:53:25 GMT+0200 | DEBUG | AuthService:refreshAuthToken | Attempting token refresh with: QfRijP...Y9xJY=
```

This clearly shows that the workaround works. There's a 10s delay between line 8 and 9 (where the delay would probably happen), the `AuthService:refreshAuthToken` is trying to refresh the token and doesn't have a null value like we used to have in pre-patch logs (indicating that storage was read successfully).

The reason `isProtectedDataAvailable` is still false here is due to the fact that I pass the value in AppDelegate's `didFinishLaunchingWithOptions`:

```objective-c
- (BOOL)application:(UIApplication *)application
    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  
  if ([FIRApp defaultApp] == nil) {
      [FIRApp configure];
  }
  
#ifdef FB_SONARKIT_ENABLED
  InitializeFlipper(application);
#endif
  
  bool isProtectedDataAvailable = [UIApplication sharedApplication].isProtectedDataAvailable;
  NSDictionary *props = @{@"isProtectedDataAvailable" : @(isProtectedDataAvailable)};

  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self
                                            launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                                   moduleName:@"ffemergency"
                                            initialProperties:props];

  if (@available(iOS 13.0, *)) {
    rootView.backgroundColor = [UIColor systemBackgroundColor];
  } else {
    rootView.backgroundColor = [UIColor whiteColor];
  }
  
  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];

  [RNBootSplash initWithStoryboard:@"BootSplash" rootView:rootView];

  return YES;
}
```

**Comment by JoniVR** [comment]
@ammarahm-ed Fair point, was just pointing out a potential issue with this but you might be right. 

> Btw are you seeing any unnecessary delays in app launch after you made the change in release?

Yes, as you can see in above logging, it does introduce startup delays, however, to me this seems like no big deal. Yes, a user will very rarely have to wait a bit longer for their app to launch, seems like an okay tradeoff vs "losing" all data, especially when you consider how relatively rarely this issue appears.

edit: 
Perhaps lowering the time between checks and increasing the amount of checks like you suggested earlier might also help a bit with this.

Anyways.. as eager as I am to release a stable version to our users with my patch, I'll wait a bit with it so I can still test any other potential solutions if you want.

**Comment by ammarahm-ed** [comment]
@JoniVR The delay can be optimized with more iterations of smaller interval, 10-15ms. Maybe you can delay the interval and test for the next few days.

**Comment by JoniVR** [comment]
@ammarahm-ed One more question, why do you think we'd need to also apply the fix to get/set/remove methods then? To me it seems like they won't need the check in that case since the data would already be available (because storage can only be initialized once it is safe).

I'll lower the interval to 25ms on my end and increase the amount of tries, don't want to poll too aggressively either.

Current setup I'll be testing:
```patch
diff --git a/node_modules/react-native-mmkv-storage/ios/SecureStorage.m b/node_modules/react-native-mmkv-storage/ios/SecureStorage.m
index eacd030..99f0859 100644
--- a/node_modules/react-native-mmkv-storage/ios/SecureStorage.m
+++ b/node_modules/react-native-mmkv-storage/ios/SecureStorage.m
@@ -5,6 +5,7 @@
 #endif
 
 #import "SecureStorage.h"
+#import <UIKit/UIKit.h>
 
 
 @implementation SecureStorage : NSObject
@@ -43,6 +44,14 @@ - (NSString *) getSecureKey:(NSString *)key
     
     @try {
         [self handleAppUninstallation];
+        dispatch_sync(dispatch_get_main_queue(), ^{
+            int readAttempts = 0;
+            // See: https://github.com/ammarahm-ed/react-native-mmkv-storage/issues/195
+            while (![[UIApplication sharedApplication] isProtectedDataAvailable] && readAttempts++ < 100) {
+                // sleep 25ms before another attempt
+                usleep(25000);
+            }
+        });
         NSString *value = [self searchKeychainCopyMatching:key];
         if (value == nil) {
             NSString* errorMessage = @"key does not present";
```

**Comment by ammarahm-ed** [comment]
@JoniVR I think you are right. Methods other than get will be called later in app lifecycle hence they don't need the above work around.

**Comment by JoniVR** [comment]
Update 1 week later: I got one new report, but that one was from a user that didn't update to my initial patched build yet. So no new reports from any of the 2 patched builds. To me this seems like a pretty clear indication the fix is working as expected and probably ready for release?

**Comment by jaltin** [comment]
@JoniVR is there a PR for this yet? Would be curious to test it too if it is available as a PR.

**Comment by JoniVR** [comment]
@jaltin No PR yet as I wasn't sure this was the solution @ammarahm-ed wanted to use. 

If you are eager to test yourself, I'd recommend looking into how to use patch files using [patch-package](https://github.com/ds300/patch-package) (pretty easy to do)

You might even be able to apply my above patch directly to the bugfix branch.

**Comment by jaltin** [comment]
Thanks for quick follow-up @JoniVR. 

Interesting how patch-package works, never seen that before. I prefer to wait for a PR, but thanks for making me aware of that package.

We have been badly affected by this bug (@andrei-tofan that initially reported this issue works with me on the same product). Yesterday we did a testflight release where we replaced the secure storage with a plain storage to see if we could avoid the bug. Today we again were logged out on on of our devices where we were testing. 

Is it possible that this bug is also applicable to the plain storage, not just the secure storage?

**Comment by JoniVR** [comment]
@jaltin That would depend on how the plain storage works, if it's using Keychain as well, it might be affected by the same bug. 

I have pretty much confirmed with certainty that this issue is related to the protected data being the culprit for our case though, so perhaps there's another issue related to it, not sure.

**Comment by ammarahm-ed** [comment]
> Thanks for quick follow-up @JoniVR. 
> 
> Interesting how patch-package works, never seen that before. I prefer to wait for a PR, but thanks for making me aware of that package.
> 
> We have been badly affected by this bug (@andrei-tofan that initially reported this issue works with me on the same product). Yesterday we did a testflight release where we replaced the secure storage with a plain storage to see if we could avoid the bug. Today we again were logged out on on of our devices where we were testing. 
> 
> Is it possible that this bug is also applicable to the plain storage, not just the secure storage?

No it's not possible. maybe your bug report is not accurate and coming from an older version of the app with secure storage.

**Comment by jaltin** [comment]
> No it's not possible. maybe your bug report is not accurate and coming from an older version of the app with secure storage.

Thanks for verifying that @ammarahm-ed. I will dig further to see if we can find another cause for that then.

**Comment by ammarahm-ed** [comment]
> Thanks for quick follow-up @JoniVR. 
> 
> Interesting how patch-package works, never seen that before. I prefer to wait for a PR, but thanks for making me aware of that package.
> 
> We have been badly affected by this bug (@andrei-tofan that initially reported this issue works with me on the same product). Yesterday we did a testflight release where we replaced the secure storage with a plain storage to see if we could avoid the bug. Today we again were logged out on on of our devices where we were testing. 
> 
> Is it possible that this bug is also applicable to the plain storage, not just the secure storage?

In your case, it might be possible but it is due to the same bug as above. That's because if you change the storage type in code but on user's device at runtime, the storage instance is encrypted, then the encrypted instance is used. Not the plain storage. Maybe that's why you are still being logged out.

**Comment by ammarahm-ed** [comment]
> > No it's not possible. maybe your bug report is not accurate and coming from an older version of the app with secure storage.
> 
> Thanks for verifying that @ammarahm-ed. I will dig further to see if we can find another cause for that then.

See my last comment.

**Comment by jaltin** [comment]
> Thanks for verifying that @ammarahm-ed. I will dig further to see if we can find another cause for that then.

An update on this. The error we saw was actually due to a coding error, so there is most likely not an issue in the non-encrypted storage.

**Comment by ammarahm-ed** [comment]
> > Thanks for verifying that @ammarahm-ed. I will dig further to see if we can find another cause for that then.
> 
> An update on this. The error we saw was actually due to a coding error, so there is most likely not an issue in the non-encrypted storage.

The above case i mentioned will still be true for users who update the app. Hence the patch-fix is the only way to properly fix this bug.

**Comment by JoniVR** [comment]
@ammarahm-ed Any idea on when you can release a version with a fix? 

You want me to submit the above patch as PR?

**Comment by ammarahm-ed** [comment]
> @ammarahm-ed Any idea on when you can release a version with a fix? 
> 
> You want me to submit the above patch as PR? 

If you can submit a PR that would be great

**Comment by JoniVR** [comment]
@ammarahm-ed Done, let me know if anything else needs to be done :)

**Comment by ammarahm-ed** [comment]
Fix shipped in **v0.7.0** Thanks @JoniVR for the insane debugging and finally fixing this too.

**Comment by JoniVR** [comment]
@ammarahm-ed 

Bad news, I would reopen this issue. I left in the automatic detection of this issue and I'm still getting this sometimes in production. This automatic detection in production also gives a better sense of the amount of times the issue was actually occurring.

I've received 18 reports from 11 different users (with around 1400 iOS users in total) just today and yesterday combined.

**Some of these reports clearly showcase the workaround working**, for example:
```log
0 | Thu Apr 21 2022 00:58:52 | DEBUG | ios | MMKV:encryptedStorage.getString | undefined
1 | Thu Apr 21 2022 00:58:52 | ERROR | ios | MMKV:encryptedStorage.getString | Expected null is not null
2 | Thu Apr 21 2022 00:58:58 | DEBUG | ios | persistErrorHandlerMiddleware | Rehydrating auth, token: eyJhbG...byZ3LY
3 | Thu Apr 21 2022 00:58:58 | DEBUG | ios | AuthService:refreshAuthToken | Attempting token refresh with: PQTUZ1...KdN9o=

``` 
Notice how there's a 6s delay between line 1 -> 2 => indicates the delay worked

**Other reports don't show a delay and also seem to still log the user out**:
```
0 | Wed Apr 20 2022 21:13:31 | DEBUG | ios | MMKV:encryptedStorage.getString | undefined
1 | Wed Apr 20 2022 21:13:31 | ERROR | ios | MMKV:encryptedStorage.getString | Expected null is not null
2 | Wed Apr 20 2022 21:13:31 | DEBUG | ios | persistErrorHandlerMiddleware | Rehydrating auth, token: null
3 | Wed Apr 20 2022 21:13:31 | DEBUG | ios | App | Notification permission status: 1
4 | Wed Apr 20 2022 21:13:31 | DEBUG | ios | setMessagingToken | OS: ios, Token: cJyUje...7-QCo4
5 | Wed Apr 20 2022 21:13:31 | DEBUG | ios | httpInterceptor:responseHandler | {
"requestUrl": "/endpoints.emergency.json",
"responseCode": 200,
"bearer": null
}
``` 

So, for some reason the delay isn't always happening.. any ideas as to why?

edit: a third kind of log:
```log
0 | Thu Apr 21 2022 14:28:47 | DEBUG | ios | MMKV:encryptedStorage.getString | undefined
1 | Thu Apr 21 2022 14:28:47 | ERROR | ios | MMKV:encryptedStorage.getString | Expected null is not null
2 | Thu Apr 21 2022 16:45:38 | DEBUG | ios | persistErrorHandlerMiddleware | Rehydrating auth, token: null
```
Here the timing between the "read" and actual re-hydration seems to be +2hrs, which I think indicates that the app launched in the background?

**Comment by zhaiyjgithub** [comment]
Still on v0.7.5.

**Comment by JoniVR** [comment]
I've created a follow-up PR, see: #263, which is related to the discussion in #246 and specifically https://github.com/ammarahm-ed/react-native-mmkv-storage/issues/246#issuecomment-1171095607

**Comment by mkhotib20** [comment]
still on 0.9.0

**Comment by scblason** [comment]
I'm on 0.8.0 (because of < 0.71 project) and also having this issue when testing on iOS simulator. Initialized shows as false when checking MMKV's options object.

**Comment by rlueder** [comment]
This is still an issue with `0.9.1`, easy to reproduce by using `setString` and `getString`, `react-native-mmkv-storage` often returns `null` after the value has supposedly been set, sometimes it'll return the value as expected, but not possible to determine what value it'll return.

After some more testing it does seem that app installs has an effect on it, if an app install is a "good" one you'll consistently get the values back as expected, if it's a "bad install" it'll return null every single time, restarting the app doesn't fix it.

I also tested side-by-side with `react-native-mmkv` and this other library consistently returns the string value as expected even after reinstalls so we're switching over to it: https://github.com/mrousavy/react-native-mmkv/

**Comment by muhammadarsal** [comment]
Hey @rlueder , we are also facing the same issue with `v0.9.1` and now migrating to react-native-mmkv. Just wanted to know if you guys were able to get rid of this issue after switching. Any heads-up would be greatly appreciated. Thanks!

**Comment by rlueder** [comment]
@muhammadarsal yes, we finished the migration a few weeks ago and the issue was fixed.

**Comment by nazmeln** [comment]
Hi @muhammadarsal,
Have you migrated to a different lib and issue is fixed? Thanks!
